inginious/frontend/task_dispensers/toc.py

```diff
@@ -100,16 +100,20 @@ def get_dispenser_data(self):
     def render_edit(self, template_helper, course, task_data):
         """ Returns the formatted task list edition form """
         config_fields = {
-            "closed": SectionConfigItem(_("Closed by default"), "checkbox", False)
+            "closed": SectionConfigItem(_("Closed by default"), "checkbox", False),
+            "hidden_if_empty": SectionConfigItem(_("Hidden if empty"),"checkbox",False)
         }
         return template_helper.render("course_admin/task_dispensers/toc.html", course=course,
                                       course_structure=self._toc, tasks=task_data, config_fields=config_fields,
                                       config_items_funcs=["dispenser_util_get_" + config_item.get_id() for config_item in self.config_items])
 
     def render(self, template_helper, course, tasks_data, tag_list):
         """ Returns the formatted task list"""
+        course_req = self._database.courses.find_one({"_id":course.get_id()})
+        people = course_req["students"] if course_req is not None else [] + course.get_staff()
+        accessibilities = course.get_task_dispenser().get_accessibilities(self._task_list_func(),people)
```

Comment:
Only the session user accessibilities have to be computed. You can also check and propagate this big dictionary accross all the templates (`task_list.html` for instance) to avoid recomputing accessibilities when printing each task.